src/persist/src/lib.rs

```diff
@@ -60,7 +59,36 @@ pub mod unreliable;
 // - Equality edge cases around all the various timestamp/frontier checks.
 
 /// A type usable as a persisted key or value.
-///
-/// TODO: Replace Abomonation with something like an Encode+Decode trait.
-pub trait Data: Clone + Abomonation + fmt::Debug + Ord {}
-impl<T: Clone + Abomonation + fmt::Debug + Ord> Data for T {}
+pub trait Data: Clone + Codec + fmt::Debug + Ord {}
+impl<T: Clone + Codec + fmt::Debug + Ord> Data for T {}
+
+/// Encoding and decoding operations for a type usable as a persisted key or
+/// value.
+pub trait Codec: Sized + 'static {
```

Comment:
this trait makes a lot of sense! do you think the trait should have something like:

```
/// The version of this codec.
fn codec_version() -> u64
```

or did you think that would be encapsulated in the name / is the wrong level of abstraction for this?

Reply:
There's two possible cases involved in bumping a version. The first is that the impl of `decode` can detect the old version and faithfully decode Self from it. This is what we'll have to do with Row, otherwise, we'll lose the ability to read data written by old versions of mz, which is a non-starter for the forseeable future. The second case is if we can't do the above, in which case, it's indistinguishable from a totally unrelated codec and we can handle that with `name` (at some point before v1.0 we'll simply RowExperimental to Row).

(Snowflake periodically rewrites every bit of data they store for encryption key migration. If we ever do the same we could potentially use that process to rip out support for old versions of the row codec. On-prem customers would then have to run the same infra. "Anything you write down is permanent technical debt" rears its ugly head again.)

I tried to pre-empt this question with some of my commentary in `decode` but given that you asked it, it seems I didn't quite pull that off. I'd like to be done rebasing this, so I'm going to merge as-is and we can discuss async how I could have made this more clear.